### ISSUE TITLE
Permanent home for the tendermint proto Dockerfile

### ISSUE THREAD

**Issue by williambanfield** [tendermint/tendermint#7272]
The Dockerfile that contains tools for building the tendermint protos lives in the tendermintdev registry, under [docker-build-proto](https://hub.docker.com/r/tendermintdev/docker-build-proto). 

The core team does not appear to have access to this account and cannot easily modify it as required. We therefore need to find a more maintainable longterm home for this Dockerfile that we have administrative permissions to access. Perhaps [tendermint](https://hub.docker.com/r/tendermint) could work, although I'm similarly uncertain if we have admin access to that.

**Comment by creachadair** [comment]
I'd advise we take a two-pronged approach here:

1. Where possible, get rid of pre-baked images from our build and CI steps, and instead build the images we need as part of the CI workflow. Judicious use of buildx caching can help reduce the expense of complex images that don't change often.
2. Where (1) is not possible, move images to ghcr.io (a.k.a. GitHub packages) and tie them to our existing GitHub permissions rather than separate credentials.

For (2), we should probably also have a separate manual workflow for updating the images. Right now, there are some places where updates to the image are gated by the same CI that requires the image to pass, and that makes updating both complicated.

To the extent possible, we should try to have a build and testing setup that is easy to replicate locally, without having to emulate GitHub Actions. It _is_ possible to run Actions locally, but it's a really heavyweight and tedious workflow, and our CI is already a meaningful impediment to velocity.

**Comment by cmwaters** [comment]
Agree strongly with a lot of what @creachadair is saying here. I guess the next question is what pre-baked images does it make sense to get rid of and what images should we instead move to ghcr.io

**Comment by creachadair** [comment]
## An informal survey of CI images based on workflow configs. This is probably incomplete.

- https://github.com/tendermint/tm-db/blob/0e6efdda1b15ef02cdb8f66b36bac942943ab7d5/.github/workflows/ci.yml#L21
- https://github.com/tendermint/spec/blob/master/.github/workflows/proto-dockerfile.yml#L32
- https://github.com/tendermint/tm-db/blob/master/.github/workflows/docker.yml#L28
- https://github.com/tendermint/images/blob/master/.github/workflows/rbuilder.yml#L16
- https://github.com/tendermint/tendermint/blob/master/.github/workflows/docker.yml#L21

## Images from DockerHub:

### [tendermintdev repository](https://hub.docker.com/search?q=tendermintdev&type=image)

Many of these do not seem to be regularly updated, and could maybe be ignored.

- tendermintdev/docker-build-proto
- tendermintdev/docker-docs-deployment
- tendermintdev/docker-hashicorp-go
- tendermintdev/docker-protoc
- tendermintdev/docker-tendermint-build
- tendermintdev/docker-tm-db-testing
- tendermintdev/docker-website-deployment
- tendermintdev/docker-website-deployment-yarn
- tendermintdev/docs_deployment
- tendermintdev/ethermint
- tendermintdev/gaia_sim
- tendermintdev/iam_management
- tendermintdev/jq_curl
- tendermintdev/rbuilder
- tendermintdev/rosetta-cli
- tendermintdev/sdk-proto-gen
- tendermintdev/tendermint
- tendermintdev/terragrunt
- tendermintdev/voyager_node
- tendermintdev/voyager_node_browser
- tendermintdev/website_deployment
- tendermintdevelopment/starport
- tendermintdevelopment/testchain

### [tendermint repository](https://hub.docker.com/search?q=tendermint%2F&type=image)

- tendermint/abacas
- tendermint/devdoc
- tendermint/ethermint
- tendermint/gaia
- tendermint/gaiadnode
- tendermint/kms
- tendermint/localnode
- tendermint/tendermint
- tendermint/testing
- tendermint/tm-signer-harness

**Comment by creachadair** [comment]
It looks like most of the rules for building these images live in the https://github.com/tendermint/images repository, but not all. Examples:

- The `tendermintdev/docker-tm-db-testing` image is built from https://github.com/tendermint/tm-db/blob/master/tools/Dockerfile.
- The `tendermint/tendermint` and `tendermint/testing` images appear to be built from https://github.com/tendermint/tendermint/tree/master/DOCKER.

While it is logical to have image definitions live in their own repository, I think we might be better off keeping image definitions in the repository that uses them. That makes it easier to figure out where to look when you need to make updates.

I think we could also probably delete some of the image definitions in the https://github.com/tendermint/images repository, several of them appear to be obsolete (e.g., the [terragrunt](https://github.com/tendermint/images/tree/master/terragrunt) image is for CircleCI, which we no longer use).

**Comment by williambanfield** [comment]
> in the repository that uses them

The immediate wrinkle to this is the proto image. That image is used across the `spec` and `tendermint` repo. 
https://github.com/tendermint/tendermint/blob/master/Makefile#L17
https://github.com/tendermint/spec/blob/master/Makefile#L1

> For (2), we should probably also have a separate manual workflow for updating the images. Right now, there are some places where updates to the image are gated by the same CI that requires the image to pass, and that makes updating both complicated.

This seems like a nice to have, although, realistically, how often are we going to updating the images at the same time as many other things? I think that for the most part, we can get away with issuing separate pull requests for image updates.

**Comment by creachadair** [comment]
> The immediate wrinkle to this is the proto image. That image is used across the `spec` and `tendermint` repo. https://github.com/tendermint/tendermint/blob/master/Makefile#L17 https://github.com/tendermint/spec/blob/master/Makefile#L1

Yes, but ideally we can pick one or the other of those repositories to host it (I would suggest the spec, since that's where the protos actually live). I think it's fine to keep shared images for workflows across multiple repositories. Rationale: Such an image would need to be independent of repo context anyway, unlike a "builder" for other kinds of code.

That said: I agree with your underlying point, and I think it's still sensible to do this for other kinds of images.

> This seems like a nice to have, although, realistically, how often are we going to updating the images at the same time as many other things? 

Probably not too often, but the problem is that when we _do_ want to do that, it's currently kind of painful. I agree this isn't worth building a complicated system for—but I don't think it's too hard to arrange (probably just "manual triggers" on the workflows would be enough).

> I think that for the most part, we can get away with issuing separate pull requests for image updates.

Agree, and I do think we should support that anyway.

**Comment by creachadair** [comment]
Since this issue is specifically about the proto builder, I assembled some of the discussion here into a new issue: #7297

**Comment by creachadair** [comment]
I think this is now resolved: Images are now pushed to ghcr.io, which we can access with our normal GitHub credentials and access permissions. Both tip and the v0.35.x backport branch use this image, which covers the active development set.

If I've missed anything, please feel free to reopen this.
